.github/workflows/validate.yml

```diff
@@ -0,0 +1,283 @@
+      # might decide to build the library but not the executable which is
+      # what we need.
+      - name: Install cabal-plan
+        run: |
+          cd $(mktemp -d)
+          cabal install cabal-plan --constraint='cabal-plan +exe'
+          echo "$HOME/.cabal/bin" >> $GITHUB_PATH
+
+      # Needed by cabal-testsuite/PackageTests/Configure/setup.test.hs
+      - name: Install Autotools
+        if: runner.os == 'macOS'
+        run: |
+          brew install automake
+
+      - name: Set validate inputs
+        run: |
+          FLAGS=""
+          if [[ "${{ matrix.cli }}" == "false" ]]; then
+            FLAGS="$FLAGS --lib-only"
+          fi
+          if [[ ${{ matrix.ghc }} == ${{ env.GHC_FOR_SOLVER_BENCHMARKS }} ]]; then
+            FLAGS="$FLAGS --solver-benchmarks"
+          fi
+          if [[ ${{ matrix.ghc }} == ${{ env.GHC_FOR_COMPLETE_HACKAGE_TESTS }} ]]; then
+            FLAGS="$FLAGS --complete-hackage-tests"
+          fi
+          echo "FLAGS=$FLAGS" >> $GITHUB_ENV
+
+      - name: Validate print-config
+        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v $FLAGS -s print-config
+
+      - name: Validate print-tool-versions
+        run: sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v $FLAGS -s print-tool-versions
+
+      - name: Validate build
+        run: |
+          sh validate.sh -j 2 -w ghc-${{ matrix.ghc }} -v $FLAGS -s build
+
+      - name: Tar cabal head executable
+        if: matrix.cli != 'false' && matrix.ghc == env.GHC_FOR_RELEASE
+        run: |
+          CABAL_EXEC=$(cabal-plan list-bin --builddir=dist-newstyle-validate-ghc-${{ matrix.ghc }} cabal-install:exe:cabal)
+          # We have to tar the executable to preserve executable permissions
+          # see https://github.com/actions/upload-artifact/issues/38
+          if [[ ${{ runner.os }} == 'Windows' ]]; then
+            CABAL_EXEC=$(cygpath $CABAL_EXEC)
+          fi
+          tar -cvf cabal-head.tar -C $(dirname "$CABAL_EXEC") $(basename "$CABAL_EXEC")
+          echo "CABAL_EXEC_TAR=cabal-head.tar" >> $GITHUB_ENV
+
+      # We upload the cabal executable built with the ghc used in the release for:
+      # - Reuse it in the dogfooding job (although we could use the cached build dir)
+      # - Make it available in the workflow to make easier testing it locally
+      - name: Upload cabal-install executable to workflow artifacts
+        if: matrix.cli != 'false' && matrix.ghc == env.GHC_FOR_RELEASE
+        uses: actions/upload-artifact@v2
+        with:
+          name: cabal-${{ runner.os }}-${{ matrix.ghc }}
+          path: ${{ env.CABAL_EXEC_TAR }}
+
+      - name: Validate lib-tests
+        env:
+          # `rawSystemStdInOut reports text decoding errors`
+          # test does not find ghc without the full path in windows
+          GHCPATH: ${{ steps.setup-haskell.outputs.ghc-exe }}
+        run: sh validate.sh -j 2 -w $GHCPATH -v $FLAGS -s lib-tests
```

Comment:
This is weird. You mean `-w ${{ matrix.ghc }}` would not work here?

Reply:
it did not work for windows in my tests, will double check and post here or openan issue about

Reply:
Just [double checked](https://github.com/jneira/cabal/runs/5553467288?check_suite_focus=true#step:13:14) ant if fails for all ghc versions with:

```
Unit Tests
  Distribution.Simple.Utils
    rawSystemStdInOut reports text decoding errors:                      FAIL
      Exception: ghc-9.2.1: createProcess: does not exist (No such file or directory)
      Use -p '/rawSystemStdInOut reports text decoding errors/' to rerun this test only.
```

Reply:
Not sure if it would be related with https://github.com/haskell/cabal/issues/7075 as the path to the ghc executable will be longer than `ghc-x.y.z` for sure

Reply:
Soooo change `-w ghc-x.y.z` to `-w ghc` seems to workaround also this one :open_mouth: 

https://github.com/jneira/cabal/runs/5553870804?check_suite_focus=true

Maybe the wrapper is interfering with the executable discovering.

So one more reason to switch to raw ghc.
Comments should be moved to validate sh though